### ISSUE TITLE
LINUX: Successfully built, OBJ import and export not working

### ISSUE THREAD

**Issue by mzogithub** [ivygen/ivygen#11]
Hi,

I built this on Ubuntu 19.04. Build was successful with no errors. The application starts normally. It seems to work fine if I generate Ivy on an empty scene. However, if I try to export a clump of ivy (which in my case it was generated without a base object), the export fails. Also, if I try to import an OBJ to generate the ivy upon, the import fails as well. For both operations I get the same error in the terminal:

`(IvyGen:24487): Gtk-WARNING **: 16:59:33.840: Failed to measure available space: Error getting filesystem info for /media/root/linuxBackup: Permission denied`

The mentioned mount point "linuxBackup" is a backup partition on a separated hard drive and should be totally unrelated to this.

When I start the applications I also get the following warnings:

`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)`
`qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)`

Thank you.

**Comment by mzogithub** [comment]
UPDATE:

The error:
`(IvyGen:24487): Gtk-WARNING **: 16:59:33.840: Failed to measure available space: Error getting filesystem info for /media/root/linuxBackup: Permission denied` goes away if I unmount that partition. However, the import/output problem persist.

Perhaps the problem is about the importer/exporter plugins not being loaded?

**Comment by mattiascibien** [comment]
Where did you put the plugin .so files?

**Comment by mattiascibien** [comment]
The warnings you see are related to OpenGL non finding that particular frame buffer config. Can you post your specs?

**Comment by mzogithub** [comment]
I’m not in front of the system right now but yes I saw the import export plugins inside the plugin folder within the ivygen-master folder and they were normal .so libraries. 

My specs are Ubuntu 19.04, graphics Nvidia 1080Ti with latest proprietary driver 430. CPU intel 9900K. 

If that helps, I had other anomalies related to OpenGL. In order to build this I had to install QT 5 dev and the open gl library that comes with it was breaking my Maya installation. In order to fix it I either had to delete QT 5 dev from my system or use a workaround and delete a open gl library that ships with Maya from the Maya lib folder. I did the latter. Not sure if this has anything to do with ivy gen though.

Also, if I generate ivy in the empty viewport everything works fine (viewport navigation etc) so open gl seems to work fine. The only two things not working are obj import and obj export. 

Thank you

**Comment by mattiascibien** [comment]
As per https://github.com/ivygen/ivygen/blob/f7ab4155cef215ee425db432682a32d0303989b3/src/app/plugin_loader.cpp#L17 I still need to implement the plugin folder for Unix systems. As a workaround you can put the .so files in a plugins folder next to the program directory and they should get loaded. I still need to figure out where to put them in the “official” Linux version as it seems I cannot find a standard that defines this (maybe /usr/share/ivygen/plugins can be good). If you know how other programs does it you are welcome to contribute a pull request.

Sorry for bad formatting but I am on my phone

**Comment by mzogithub** [comment]
Thanks. I moved libobjexporter.so and libobjimporter.so into the bin folder where the IvyGen executable is located. Before they  were in a plugins folder and there they seemed to be ignored.

Now the import works. However, the export still does not.  I don't get any specific error in the terminal when I try to export. I only get the mount error I mentioned before that is probably unrelated.

This may still be a path problem or maybe something wrong in obj_writer.cpp?

**Comment by mzogithub** [comment]
UPDATE: now the export works but I had to copy both the libraries in both ./bin and in a new folder ./bin/plugins. For some strange reason they needed to be copied in both locations otherwise export does not work. I guess this is just a matter of fixing the paths in plugin_loader.cpp as you mentioned. 

Regarding where to put those libraries in the official Linux version, I'm not an expert but normally the libraries shipping with an application are located in a lib folder within the app root folder. If you have an installer, sometimes symbolic links to system libraries located in /lib/x86_64-linux-gnu are created instead. That being said, for this small applications I believe the best way would be to pack it into an Appimage which contains everything the app needs to work.

Thank you

**Comment by mattiascibien** [comment]
The exporter in both locations seems strange to me. At the moment I do not have access to a PC but as soon as I get access to a Linux machine I’ll submit a correction.

**Comment by mzogithub** [comment]
Yes, it seems strange to me as well. However, I tried all the possible combinations and discovered that accidentally. In fact, I tried with both libs inside ./bin and only import worked. Then I copied both of them into ./bin/plugins but I forgot to delete them from ./bin and both worked. When I went to clean up and deleted the duplicates from ./bin, again only import was working. So I copied them in both places and now input and export both work. I really don’t know why.

**Comment by mattiascibien** [comment]
Thanks for all the info you are providing me. As soon as I get access to a Linux machine I’ll put plugins in the right directory and release a new version (maybe a app image and Flatpak/snap).

**Comment by mattiascibien** [comment]
@mzogithub I have built the software on my Linux machine and I can confirm the behavior you are facing. Still not sure about the issue but I'll try to fix it and maybe release a 1.0 version.

**Comment by mzogithub** [comment]
@mattiascibien Sounds good, thanks for letting me know.

**Comment by mattiascibien** [comment]
@mzogithub I am going to do a refactor of the whole plugin thing (was a bit of a overkill for this kind of software) and i guess that for the next version I am going to implement model loading using [Assimp](http://www.assimp.org/) so I can handle multiple model formats without too much hassle.

Not sure how this is going to work but when it is completed I'll try to push some binaries (0.11.0) even for linux so you probably won't need to compile again. 

The first format will probably be AppImage.
